### PR TITLE
add ``by_alias`` argument in ``.dict()`` and ``.json()`` model methods

### PR DESCRIPTION
**Pull request by Gr1N**
#205

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -181,22 +181,26 @@ def __getstate__(self):
...
         object.__setattr__(self, '__values__', state)
 
-    def dict(self, *, include: Set[str]=None, exclude: Set[str]=set()) -> Dict[str, Any]:
+    def dict(self, *, include: Set[str]=None, exclude: Set[str]=set(), by_alias: bool = False) -> Dict[str, Any]:
         """
         Generate a dictionary representation of the model, optionally specifying which fields to include or exclude.
         """
         return {
-            k: v for k, v in self
+            self.fields[k].alias if by_alias else k: v
```
Comment (samuelcolvin) [inline pydantic/main.py]: This is called a lot so should be as efficient as possible, I guess

```py
if by_alias:
   return ...
else:
   return ...
```

File: pydantic/main.py [hunk 2]
```diff
@@ -181,22 +181,26 @@ def __getstate__(self):
     def __setstate__(self, state):
         object.__setattr__(self, '__values__', state)
 
-    def dict(self, *, include: Set[str]=None, exclude: Set[str]=set()) -> Dict[str, Any]:
+    def dict(self, *, include: Set[str]=None, exclude: Set[str]=set(), by_alias: bool = False) -> Dict[str, Any]:
```
Comment (samuelcolvin) [inline pydantic/main.py]: I guess `by_alias` needs to be recursive (eg. work on submodels).

To do this:
* rename `__iter__()` to `_iter(by_alias=False)`
* `__iter__()` can then just call `_iter()`
* `_get_value` should have signature `_get_value(by_alias)` and then pass `by_alias` on.

File: tests/test_main.py [hunk 3]
```diff
@@ -298,6 +299,7 @@ class Config:
 
     assert Model(a='different').a == 'different'
     assert Model(a='different').dict() == {'a': 'different'}
+    assert Model(a='different').dict(by_alias=True) == {'_a': 'different'}
```
Comment (samuelcolvin) [inline tests/test_main.py]: need a test for submodels.

Comment (Gr1N) [reply tests/test_main.py]: added

File: pydantic/main.py (hunk 2 above)
Comment (Gr1N) [reply pydantic/main.py]: updated

**Pull request merged**